### PR TITLE
Support for Jetpack site icons

### LINKED ISSUE #3720: Jetpack blavatars
In My Sites & Blog detail pages the custom site icon selected in the Jetpack sites are not showing up currently. It's due to constructing a gravatar url for the site address and not using what the REST api returns for the site icon.

### PR DESCRIPTION
**Pull request by oguzkocer**
Closes #3720. The issue we wanted to fix was the Jetpack site icons were not showing up in the app. Up until now, we used the host url to construct a gravatar.com url for the blavatar. However, we do not support this for Jetpack sites. In order to solve the issue, we need to start saving the icon url for the `Blog`. As discussed on Slack with @koke, we have opted to use the term "icon" instead of "blavatarUrl" since the url returned might not be a blavatar url. Also, we originally wanted to move the URL construction away from `UIImageView+Gravatar` since it's a UIKit category, but we need the size information, so it's probably best to keep it there.

Needs Review: @koke (Thank you!)

### CODE REVIEW

File: WordPress/Classes/ViewRelated/NUX/AddUsersBlogCell.m [hunk 1]
```diff
@@ -78,7 +78,7 @@ - (void)layoutSubviews
     y = (rowHeight - cellBlavatarSide)/2.0;
     _blavatarImage.frame = CGRectIntegral(CGRectMake(x, y, cellBlavatarSide, cellBlavatarSide));
     NSURL *blogURL = [NSURL URLWithString:_blavatarUrl];
-    [_blavatarImage setImageWithBlavatarUrl:[blogURL host]];
+    [_blavatarImage setImageWithSiteIcon:[blogURL host]];
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/AddUsersBlogCell.m]: See #3703 :wink:

File: WordPress/Classes/Models/Blog.m [hunk 2]
```diff
@@ -101,18 +99,22 @@ - (void)setGeolocationEnabled:(BOOL)value
...
+        return icon;
     }
 
-    return _blavatarUrl;
+    // if the icon is not set we can use the host url to construct it
+    NSString *hostUrl = [[NSURL URLWithString:self.xmlrpc] host];
+    if (hostUrl == nil) {
+        hostUrl = self.xmlrpc;
+    }
+    return hostUrl;
```
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: Do we still need to fallback to the host? And if we don't, can we make `blog.icon` a `NSURL`?

File: WordPress/Classes/Categories/UIImageView+Gravatar.m [hunk 3]
```diff
@@ -50,19 +50,21 @@ - (void)setImageWithGravatarEmail:(NSString *)emailAddress fallbackImage:(UIImag
...
 {
-    if ([blavatarUrl rangeOfString:@"gravatar.com/blavatar"].location == NSNotFound) {
-        [self setImageWithURL:[self blavatarURLForHost:blavatarUrl] placeholderImage:placeholderImage];
+    if ([self isPhotonURL:siteIcon]) {
+        [self setImageWithURL:[self siteIconURLForSiteIconUrl:siteIcon] placeholderImage:placeholderImage];
+    } else if ([self isBlavatarURL:siteIcon]) {
+        [self setImageWithURL:[self blavatarURLForBlavatarURL:siteIcon] placeholderImage:placeholderImage];
     } else {
-        [self setImageWithURL:[self blavatarURLForBlavatarURL:blavatarUrl] placeholderImage:placeholderImage];
+        [self setImageWithURL:[self blavatarURLForHost:siteIcon] placeholderImage:placeholderImage];
```
Comment (koke) [inline WordPress/Classes/Categories/UIImageView+Gravatar.m]: I see a couple places in the reader that still pass a hostname to this method. It's probably better to leave this for a separate PR, but could we get the icon URL from the REST response?

File: WordPress/Classes/Models/Blog.m (hunk 2 above)
Comment (oguzkocer) [inline WordPress/Classes/Models/Blog.m]: I am not really sure if we need the host anymore. But I think we need to keep it as a string because Photon and gravatar sizing parameters are different. See: https://github.com/wordpress-mobile/WordPress-iOS/blob/issue/3720-jetpack-blavatars/WordPress/Classes/Categories/UIImageView+Gravatar.m#L89 So, it probably should stay as a string.

File: WordPress/Classes/Categories/UIImageView+Gravatar.m (hunk 3 above)
Comment (oguzkocer) [inline WordPress/Classes/Categories/UIImageView+Gravatar.m]: Are you talking about this bit: https://github.com/wordpress-mobile/WordPress-iOS/blob/issue/3720-jetpack-blavatars/WordPress/Classes/ViewRelated/Reader/ReaderBrowseSiteViewController.m#L100-L104? If so, I don't think the REST response has it, I might be looking into the wrong place though: https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/

File: WordPress/Classes/ViewRelated/NUX/AddUsersBlogCell.m (hunk 1 above)
Comment (oguzkocer) [inline WordPress/Classes/ViewRelated/NUX/AddUsersBlogCell.m]: I was wondering about the same thing. Let's separate that to a new PR, once this is merged in I can take care of that :)

File: WordPress/Classes/Categories/UIImageView+Gravatar.m (hunk 3 above)
Comment (koke) [inline WordPress/Classes/Categories/UIImageView+Gravatar.m]: More like https://developer.wordpress.com/docs/api/1.1/get/read/following/, but I don't see the icon in the response. Let's leave it for the next PR :wink:

**Comment by oguzkocer** [comment]
@koke: I think there is only one comment left that needs your attention. Do you mind taking a second look at the PR when you have time? Once this PR gets merged, I'll take care of #3703.

File: WordPress/Classes/Models/Blog.m (hunk 2 above)
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: Let's leave the NSURL conversion for another time, but I think `-[Blog icon]` shouldn't return the host name. Even if the return value is a `NSString`, it should always be a valid known URL or `nil`

File: WordPress/Classes/Categories/UIImageView+Gravatar.m [hunk 4]
```diff
@@ -81,6 +83,13 @@ - (NSURL *)gravatarURLForEmail:(NSString *)email withSize:(NSInteger)size gravat
     return [NSURL URLWithString:gravatarUrl];
 }
 
+- (NSURL *)siteIconURLForSiteIconUrl:(NSString *)path
+{
+    NSInteger size = [self sizeForBlavatarDownload];
+    NSString *siteIconUrl = [NSString stringWithFormat:@"%@?w=%d&h=%d", path, size, size];
```
Comment (koke) [inline WordPress/Classes/Categories/UIImageView+Gravatar.m]: The blavatar code already had this problem, but what happens if the photon URL already has sizing parameters?

File: WordPress/Classes/Models/Blog.m (hunk 2 above)
Comment (oguzkocer) [inline WordPress/Classes/Models/Blog.m]: I agree with that. I might come back to it once I implement it in Android side as well and have a better solution for it.

File: WordPress/Classes/Categories/UIImageView+Gravatar.m (hunk 4 above)
Comment (oguzkocer) [inline WordPress/Classes/Categories/UIImageView+Gravatar.m]: Hmm... Is that even a possibility, assuming there is no developer errors?

I guess we can add some kind of string check, but I am not sure if that'd be a healthy approach. What do you think?

Comment (koke) [inline WordPress/Classes/Categories/UIImageView+Gravatar.m]: The URL comes from the API, right? Even though I don't see any parameters in actual responses, I've seen code that sets `w=80`
https://github.com/Automattic/jetpack/blob/d5f87714a05fb3588dc8809600c29b5005abaadc/json-endpoints/class.wpcom-json-api-get-site-endpoint.php#L151

`NSURLComponents` might help in parsing the URL and checking for existing query parameters, and replacing them if they exist.

**Comment by koke** [comment]
Looks great :shipit:

**Comment by oguzkocer** [comment]
Awesome, thanks for the review!
